efi/preinstall/profile.go

```diff
@@ -0,0 +1,339 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2024 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package preinstall
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+
+	"github.com/canonical/go-tpm2"
+	secboot_efi "github.com/snapcore/secboot/efi"
+	internal_efi "github.com/snapcore/secboot/internal/efi"
+)
+
+type authorityTrust int
+
+const (
+	authorityTrustBootCode authorityTrust = 1 << iota // authority is trusted to load boot code (we don't need PCR4)
+	authorityTrustDrivers                             // authority is trusted to load drivers (we may not need PCR2)
+)
+
+type authorityTrustDataSet []authorityTrustData
+
+func (s authorityTrustDataSet) determineTrust(certs []*X509CertificateID) authorityTrust {
+	trust := authorityTrustBootCode | authorityTrustDrivers
+	for _, cert := range certs {
+		var certTrust authorityTrust
+		for _, auth := range s {
+			if !bytes.Equal(auth.Authority.Subject, cert.RawSubject()) {
+				continue
+			}
+			if !bytes.Equal(auth.Authority.SubjectKeyId, cert.SubjectKeyId()) {
+				continue
+			}
+			if auth.Authority.PublicKeyAlgorithm != cert.PublicKeyAlgorithm() {
+				continue
+			}
+			if !bytes.Equal(auth.Authority.Issuer, cert.RawIssuer()) {
+				continue
+			}
+			if !bytes.Equal(auth.Authority.AuthorityKeyId, cert.AuthorityKeyId()) {
+				continue
+			}
+			if auth.Authority.SignatureAlgorithm != cert.SignatureAlgorithm() {
+				continue
+			}
+			certTrust = auth.Trust
+			break
+		}
+		trust &= certTrust
+	}
+
+	return trust
+}
+
+func (s authorityTrustDataSet) trustedForBootManager(certs []*X509CertificateID) bool {
+	return s.determineTrust(certs)&authorityTrustBootCode > 0
+}
+
+func (s authorityTrustDataSet) trustedForDrivers(certs []*X509CertificateID) bool {
+	return s.determineTrust(certs)&authorityTrustDrivers > 0
+}
+
+type authorityTrustData struct {
+	Authority *internal_efi.SecureBootAuthorityIdentity
+	Trust     authorityTrust
+}
+
+var (
+	knownCAs = authorityTrustDataSet{
+		{internal_efi.MSUefiCA2011, 0},
+		{internal_efi.MSUefiCA2023, 0}, // be conservative here for now, but will we be able to set the authorityTrustDrivers flag for the MS2023 CA?
```

Comment:
You understood correctly - these are both explicitly distrusted, as is any certificate that doesn't appear in this list. Eventually we'll get to the point where we want security features that will require our own CA, which will be added here with both trust flags set (because we know we'll never use it to sign drivers). It also remains to be seen whether Microsoft will be any more open wrt to their signing requirements with the new UEFI CA, particularly wrt the signing of option ROMs. If this improves, we might be able to change one flag for this CA, but right now, I don't think it's been used to sign anything.